### PR TITLE
Add billing module

### PR DESCRIPTION
**Pull request by jorg-vr**
This pull request adds a new module type 'billing'.

It has a single endpoint that supports a post request with contract info and meter data.
The contract info can be specified using the contract info provided by the members module.

It returns a detailed costs breakdown.

Screen Shot 2026-04-14 at 13 13 51

This pr also adds a debug mode that gives request headers with profiling info, which can be used to profile slow requests.

It also updates the members module. Members modules now return contract info, that can be used with the billing module. The example data for both the members module and the billing module is synced so output of the members module can be used as input of the billing module.

### CODE REVIEW

File: src/cofy/modules/billing/models/billing_request.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        ]
+    )
+
+    def to_meter(self) -> Meter:
+        return Meter(
+            direction=self.direction,
+            type=self.type,
+            data=pd.DataFrame(
+                {
+                    "timestamp": [dp.timestamp for dp in self.data],
```
Comment (JrtPec) [inline src/cofy/modules/billing/models/billing_request.py]: When timestamps cross the DST boundary, for example:
```python
{"timestamp": "2024-03-31T01:45:00+01:00", "value": 1.0},
{"timestamp": "2024-03-31T03:00:00+02:00", "value": 2.0}
```
the dataframe constructor does not cast them to a proper datetime dtype, which breaks code down the line.

I suggest to do `timestamps = pd.to_datetime([dp.timestamp for dp in self.data], utc=True).tz_convert(timezone)`. Side effect, I fear, is that you must ask the calling code to specify the timezone to work in.

File: src/cofy/modules/billing/models/billing_request.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+class DataPoint(BaseModel):
+    timestamp: dt.datetime = Field(examples=["2024-01-01T00:00:00+01:00"])
+    value: float = Field(examples=[150.5])
+
+
+class MeterInfo(BaseModel):
+    direction: PowerDirection = Field(default=PowerDirection.CONSUMPTION, examples=["consumption"])
+    type: MeterType = Field(default=MeterType.SINGLE_RATE, examples=["single_rate"])
+    data: list[DataPoint] = Field(
```
Comment (JrtPec) [inline src/cofy/modules/billing/models/billing_request.py:21]: Passing an empty list here breaks (500), should be caught by validation and return a 422.

File: src/cofy/modules/billing/models/billing_request.py (hunk 1 above)
Comment (JrtPec) [reply src/cofy/modules/billing/models/billing_request.py]: I guess `timezone` could come from potentially multiple places:

- The calling code, they know what timezone the input data is in.
- The contract info, if we assume that a certain contract is always evaluated in a certain timezone context
- Module config, if we assume that a COFYcloud will be run within a specific timezone

**Review by JrtPec — CHANGES_REQUESTED** [review]

File: src/cofy/modules/billing/models/billing_request.py (hunk 1 above)
Comment (jorg-vr) [reply src/cofy/modules/billing/models/billing_request.py]: Thanks for the good edge case test. I've already tried to resolve multiple timezone issues, but I didn't take into account multiple timezones in the dataframe.

We already handle different timezones between dataframe and start and end dates, and different timezones used in the indexes, or the formula definitions.
But in all those cases I just picked one 'preferred' timezone implicitly and converted all others to that timezone. (The one I picked was the one from the dataframe, thus the dataframe was the only thing not itself going through timezone conversion)

I think it makes sense to move to explicit timezone specification.
The correct place would be the contract: to correctly define a contract you already need to know the location (eg to find the distributor and tax specification), you can thus also know the timezone.
This timezone is then also very likely to line up with the timezone of the index, provider, tax changes etc,... avoiding edge case that could occur when trying to calculate billing Flemish tariffs based on Australian time when the end user is on vacation for example.

I'll update this in energy-cost

For now we can set the the timezone fixed in the cofy cloud contract definition (just like we set taxes and fees). In the future we can add a location based lookup.

File: src/cofy/api/debug_middleware.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        request_id = str(uuid.uuid4())
+        t_start = time.perf_counter()
+        request_dir = self._debug_dir / request_id
+        request_dir.mkdir(parents=True, exist_ok=True)
+
+        import yappi  # noqa: PLC0415
+
+        yappi.set_clock_type("wall")
+        yappi.start(builtins=False)
```
Comment (JrtPec) [inline src/cofy/api/debug_middleware.py]: Minor issue here: yappi start, stop and clear_stats are process wide, so you cannot debug concurrent requests. Calling a second request while the first is still running would corrupt the data. Could be helped by adding an asyncio.Lock?

File: src/cofy/api/debug_router.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+from fastapi.responses import PlainTextResponse
+
+
+class DebugRouter(APIRouter):
+    """Router exposing debug information captured by DebugMiddleware."""
+
+    def __init__(self, debug_dir: Path) -> None:
+        super().__init__(prefix="/debug", tags=["debug"])
+        self._debug_dir = debug_dir
+        self.add_api_route("/{request_id}/profile", self._get_profile, methods=["GET"])
```
Comment (JrtPec) [inline src/cofy/api/debug_router.py:13]: Debug mode has no auth? Could be an issue if debug mode is accidentally left on in a production environment.

File: src/cofy/modules/billing/models/billing_request.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            data.distributors[distributor] if distributor and distributor in data.distributors else None
+        )
+        product_tariff = products_for_ct[product] if product and product in products_for_ct else None
+
+        return Contract(
+            supplier=product_tariff,
+            distributor=distributor_tariff,
+            fees=data.fees[self.customer_type],
+            taxes=data.taxes,
+            timezone=ZoneInfo("Europe/Brussels"),
```
Comment (JrtPec) [inline src/cofy/modules/billing/models/billing_request.py:95]: Hardcoded timezone? Good for now but consider having this as a parameter per contract.

**Review by JrtPec — APPROVED** [review]

File: src/cofy/api/debug_router.py (hunk 4 above)
Comment (jorg-vr) [reply src/cofy/api/debug_router.py:13]: debug mode has auth (Auth is added to the whole api, also docs etc,...) only `/health` has no auth. This is not very explicit but it directly uses the `add_route` method instead of `add_api_route`, which does not make use of middleware (and security is a middleware layer)

`add_api_route` does use all the default middleware, thus it is also affected by our security.
